lib/services/jira/jira_wiki_converter.rb

```diff
@@ -0,0 +1,322 @@
+require "reverse_markdown"
+
+module AhaServices
+  class JiraWikiConverter
+
+    def convert_html_from_aha(html)
+      preprocess html
+      wiki = converter.convert(html)
+      postprocess wiki
+    end
+
+    def preprocess(html)
+      # replace special entities.
+      html.gsub!(/&(mdash|#8212);/, '---')
+      html.gsub!(/&(ndash|#8211);/, '--')
+      # replace non-breaking spaces
+      html.gsub!(/&(nbsp|#160);/, ' ')
+    end
+
+    def postprocess(wiki)
+      wiki.strip
+    end
+
+    def converter
+      ReverseMarkdown.new(unknown_tags: :bypass) do |c|
+        c.register(:hr, Hr.new)
+        c.register(:li, Li.new)
+        c.register(:a, A.new)
+        c.register(:img, Img.new)
+        c.register(:pre, Pre.new)
+        c.register(:code, Code.new)
+        c.register(:blockquote, Blockquote.new)
+        c.register(:font, Font.new)
+        c.register(:tr, Tr.new)
+        c.register(:td, Td.new)
+        c.register(:th, Th.new)
+
+        register_mark(c, [:strong, :b], "*")
+        register_mark(c, [:em, :i], "_")
+        register_mark(c, [:del, :strike, :s], "-")
+        register_mark(c, [:u, :ins], "+")
+        register_mark(c, [:cite], "??")
+        register_mark(c, [:sup], "^")
+        register_mark(c, [:sub], "~")
+
+        6.times do |n|
+          c.register(:"h#{n + 1}", H.new(n + 1))
+        end
+      end
+    end
+
+    class Mark < ReverseMarkdown::Converters::Base
+      def initialize(tags, surround_with)
+        @tags = tags
+        @surround_with = surround_with
+        super()
+      end
+
+      def convert(node, index)
+        content = treat_children(node)
+        if content.strip.empty? || already_marked?(node)
+          content
+        else
+          # Jira cannot handle the marks being against whitespace like * bold *
+          # so this moves the whitespace out of the mark.
+          pre_content = ""
+          post_content = ""
+
+          if content =~ /^\s+/
```

Comment:
Why use `^` and `\z`? Shouldn't you pair `^` with `$` or `\A` with `\z`? I don't think it matters here, but it seems inconsistent to me.

Reply:
Yea, this happened because I did some benchmarking and found matching $ to be a potential performance issue with large documents. But there's no difference between ^ and \A in this situation in performance or matching characteristic.